proxy/coco/src/project/create.rs

```diff
@@ -75,16 +93,61 @@ impl<Path: AsRef<path::Path>> Repo<Path> {
     pub fn project_name(&self) -> Result<String, Error> {
         match self {
             Self::Existing { path } => path
-                .as_ref()
                 .components()
                 .next_back()
                 .and_then(|component| component.as_os_str().to_str())
                 .map(ToString::to_string)
-                .ok_or_else(|| Error::EmptyExistingPath(path.as_ref().to_path_buf())),
+                .ok_or_else(|| Error::EmptyExistingPath(path.to_path_buf())),
             Self::New { name, .. } => Ok(name.to_string()),
         }
     }
 
+    /// Validate `Repo` into a [`ValidatedRepo`], ensuring that path passes certain criteria. These
+    /// criteria are documented in the **Errors** section.
+    ///
+    /// # Errors
+    ///
+    ///   * If we have `Repo::Existing` and the path does not exist
+    ///   * If we have `Repo::Existing` and the path does not point to a repository
+    ///   * If we have `Repo::New` and the path already exists
+    pub fn validate(self) -> Result<ValidatedRepo, validation::Error> {
+        match self {
+            Self::Existing { path } => {
+                if !path.exists() {
+                    return Err(validation::Error::PathDoesNotExist(path));
+                }
+
+                // TODO(finto): This may be too permissive. Need to look into git errors.
+                if git2::Repository::open(path.clone()).is_err() {
+                    return Err(validation::Error::NotARepo(path));
+                }
+
+                Ok(ValidatedRepo(Self::Existing { path }))
+            },
+            Self::New { name, path } => {
+                let repo_path = path.join(name.clone());
+                if repo_path.exists() {
```

Comment:
Is this not too strict? An empty directory should be valid for a new repo, or am I misunderstanding?

Reply:
Good shout :+1: